ArduPilot/ardupilot
# AP_DDS: External Control enable

# User Story

As an operator, I want to be able to completely cut off any external control from the DDS interface. This functionality is especially helpful during debugging tests, when a quick reaction is of utmost importance.
I would like to be able to assign a switch to Enable/Disable external control, as well as the ability for a script or an external GCS to trigger External Control.

# What Changed

- AP_ExternalControl has now the ability to set and retrieve its status.
- AP_DDS prevents any Subscription or Service to affect the vehicle: Subscriptions callbacks are simply returned, while service requests reply with `false`.
- RC_Channel has been given a new functionality `DDS_EXTERNAL_CONTROL = 181` which allows a channel to enable External control when HIGH.

# Test

- Start SITL (plane or copter)
- Call a mode switch or arm service call and verify that the call is successful.
```
ros2 service call /ap/mode_switch ardupilot_msgs/srv/ModeSwitch "mode: <mode_num>"
ros2 service call /ap/arm_motors  ardupilot_msgs/srv/ArmMotors "arm: true"
```
- Set the RC switch to High `ros2 topic pub /ap/joy sensor_msgs/msg/Joy "{axes: [0.0, 0.0, 0.0, 0.0, .nan, 1]}"`
- Assigned RC6_OPTION = 181
- Notice that the GCS will show `Command rejected: External Control Enabled`.
- Call other services and verify they are responded to succesfully.
- Set the RC switch to Low `ros2 topic pub /ap/joy sensor_msgs/msg/Joy "{axes: [0.0, 0.0, .nan, .nan, .nan, -1]}"`
- Try to Arm the vehicle with `ros2 service call /ap/arm_motors ardupilot_msgs/srv/ArmMotors "arm: true"`
- Arming command is rejected!

![image](https://github.com/user-attachments/assets/f5000107-b26e-48c2-ba37-8565add0c51c)

## Test with Mavproxy

Disable all the channels first and then set the rc channel dedicated to enabling the external control high.

```bash
rc all 65535
rc 6 1900
```
